src/renderer/base/renderer.cpp

```diff
@@ -64,44 +64,90 @@ Renderer::~Renderer()
+        }
+
+        LOG_HR_IF(hr, hr != E_PENDING);
 
-            const auto hr = _PaintFrameForEngine(pEngine);
-            if (SUCCEEDED(hr))
+        if (--tries == 0)
+        {
+            // Stop trying.
+            _pThread->DisablePainting();
+            if (_pfnRendererEnteredErrorState)
             {
-                break;
+                _pfnRendererEnteredErrorState();
             }
+            // If there's no callback, we still don't want to FAIL_FAST: the renderer going black
+            // isn't near as bad as the entire application aborting. We're a component. We shouldn't
+            // abort applications that host us.
+            return S_FALSE;
+        }
+
+        // Add a bit of backoff.
+        // Sleep 150ms, 300ms, 450ms before failing out and disabling the renderer.
+        Sleep(renderBackoffBaseTimeMilliseconds * (maxRetriesForRenderEngine - tries));
+    }
 
-            LOG_HR_IF(hr, hr != E_PENDING);
+    return S_OK;
+}
 
-            if (--tries == 0)
+[[nodiscard]] HRESULT Renderer::_PaintFrame() noexcept
+{
+    {
+        _pData->LockConsole();
+        auto unlock = wil::scope_exit([&]() {
+            _pData->UnlockConsole();
+        });
+
+        // Last chance check if anything scrolled without an explicit invalidate notification since the last frame.
+        _CheckViewportAndScroll();
+
+        if (_currentCursorOptions)
+        {
+            const auto coord = _currentCursorOptions->coordCursor;
+            const auto& buffer = _pData->GetTextBuffer();
+            const auto lineRendition = buffer.GetLineRendition(coord.y);
+            const auto cursorWidth = _pData->IsCursorDoubleWidth() ? 2 : 1;
+
+            til::rect cursorRect{ coord.x, coord.y, coord.x + cursorWidth, coord.y + 1 };
+            cursorRect = BufferToScreenLine(cursorRect, lineRendition);
+
+            if (buffer.GetSize().TrimToViewport(&cursorRect))
             {
-                // Stop trying.
-                _pThread->DisablePainting();
-                if (_pfnRendererEnteredErrorState)
+                FOREACH_ENGINE(pEngine)
                 {
-                    _pfnRendererEnteredErrorState();
+                    LOG_IF_FAILED(pEngine->InvalidateCursor(&cursorRect));
                 }
-                // If there's no callback, we still don't want to FAIL_FAST: the renderer going black
-                // isn't near as bad as the entire application aborting. We're a component. We shouldn't
-                // abort applications that host us.
-                return S_FALSE;
             }
+        }
 
-            // Add a bit of backoff.
-            // Sleep 150ms, 300ms, 450ms before failing out and disabling the renderer.
-            Sleep(renderBackoffBaseTimeMilliseconds * (maxRetriesForRenderEngine - tries));
+        _currentCursorOptions = _GetCursorInfo();
+
+        FOREACH_ENGINE(pEngine)
+        {
+            RETURN_IF_FAILED(_PaintFrameForEngine(pEngine));

```

Comment:
... and then the potential issue is that if the first engine fails out, the second engine doesn't get updated cursor info here. 


Yea that seems kinda fine. You're already in the state where at least one engine is being goofy. Both are going to get fully invalidated when we recover, seems not that bad to me 🤷